permission-element.bs

```diff
@@ -19,7 +18,408 @@ Abstract: A `<permission>` HTML element to request browser permissions in-page.
+    [=current high resolution time=].
+
+</div>
+
+NOTE: The spec does not define a way for expiring blockers that have expired
+    to be removed, so {{[[BlockerList]]}} may grow indefinitely.
+    Since the presence of expired blockers is not observable, user agents
+    should feel free to clean up the blocker list.
+
+## <{permission}> element algorithms ## {#permission-element-algorithms}
+
+<div algorithm="HTMLPermissionElement/constructor">
+The {{HTMLPermissionElement}} constructor steps are:
+
+1. Initialize the internal {{[[Types]]}} slot to null.
+1. Initialize the internal {{[[BlockerList]]}} to &laquo;[]&raquo;.
+
+</div>
+
+<div algorithm="HTMLPermissionElement/insertion steps">
+The {{HTMLPermissionElement}} [=insertion steps=] are:
+
+1. If {{[[Types]]}} is null, set {{[[Types]]}} to &laquo;[]&raquo;.
+1. Initialize the internal {{[[BlockerList]]}} to &laquo;[]&raquo;
+1. Initialize the internal {{[[IntersectionObserver]]}} with the result of
+    constructing a new {{IntersectionObserver}}, with
+    [=HTMLPermissionElement/IntersectionObserver callback=].
+1. Call {{[[IntersectionObserver]]}}.observe([=this=]).
+1. [=Add an expiring blocker=] with reason `"recently_created"`.
+
+</div>
+
+<div algorithm="HTMLPermissionElement/isValid">
+HTMLPermissionElement |element|'s <dfn>isValid</dfn> getter steps are:
+
+1. Return whether |element|'s [=HTMLPermissionElement/blocker=] is Nothing.
+
+</div>
+
+<div algorithm="HTMLPermissionElement/invalidReason">
+HTMLPermissionElement |element|'s <dfn>invalidReason</dfn> getter steps are:
+
+1. If |element| {{HTMLPermissionElement/isValid}}, then return
+    [=HTMLPermissionElement/blocker=]'s reason string.
+1. Otherwise, return `""`.
+
+</div>
+
+<div algorithm=activation>
+A <{permission}> |element|'s [=EventTarget/activation behavior=] given |event| is:
+
+1. [=Assert=]: |element|'s {{[[Types]]}} is not null.
+1. If |element|'s {{[[Types]]}} is [=list/empty=], then return.
+1. If |event|'s {{Event/isTrusted}} attribute is not true, then return.
+1. If |element|'s {{HTMLPermissionElement/isValid}} attribute is false,
+    then return.
+1. [=Request permission to use=] the powerful features named in |element|'s
+    {{[[Types]]}}.
+
+Issue: What about event handlers?
+</div>
+
+<div algorithm="HTMLPermissionElement/IntersectionObserver callback">
+The HTMLPermissionElement's <dfn for="HTMLPermissionElement">IntersectionObserver callback</dfn> implements {{IntersectionObserverCallback}} and runs the following steps:
+
+1. [=Assert=]: The {{IntersectionObserver}}'s {{IntersectionObserver/root}} is the [=Document=]
+1. Let |entries| be the value of the first callback parameter,
+    the [=/list=] of {{IntersectionObserverEntry|intersection observer entries}}.
+1. Initialize |covered| with false.
+1. Initialize |visible| with true.
+1. [=list/iterate|For any=] |item| in |entries|:
+    1. If |item|[{{IntersectionObserverEntry/isVisible}}] is false: Set |covered| to true.
```

Comment:
I think this algorithm is wrong, what we intend to do here is to 
1) either add a blocker with reason "intersection_occluded_or_distorted" or "intersection_out_of_viewport_or_clipped" (but not both) if the necessary conditions are met
2) otherwise if the visibility has changed since the last time (so we'll probably need an additional internal state field to hold this) and the element has transitioned from not visible to visible, to remove the "intersection_occluded_or_distorted"/"intersection_out_of_viewport_or_clipped" blockers and instead add an expiring "intersection_visible" blocker.

Reply:
Done. (Based on your proposal, offline.)